cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,456 @@
+			if want, got := test.expectedError, err; want != got {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+
+			if diff := cmp.Diff(test.expectedFirewall, fw); diff != "" {
+				t.Errorf("Get() mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestFirewallController_Set(t *testing.T) {
+	testcases := []struct {
+		name                             string
+		fwCache                          firewallCache
+		inboundRules                     []godo.InboundRule
+		expectedGodoFirewallAddRulesResp func(context.Context, string, *godo.FirewallRulesRequest) (*godo.Response, error)
+		expectedGodoFirewallCreateResp   func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallGetResp      func(context.Context, string) (*godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallListResp     func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error)
+		expectedGodoFirewallUpdateResp   func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error)
+		expectedError                    error
+		expectedFirewall                 *godo.Firewall
+	}{
+		{
+			name:    "create firewall when cache does not exist (i.e. initial startup)",
+			fwCache: newFakeFirewallCacheEmpty(),
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{}, newFakeNotFoundResponse(), nil
+			},
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(fakeWorkerFWName, fakeInboundRule), newFakeOKResponse(), nil
+			},
+		},
+		{
+			name:         "failing to update the firewall because it is missing",
+			fwCache:      newFakeFirewallCache(diffFakeInboundRule),
+			inboundRules: []godo.InboundRule{fakeInboundRule},
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errors.New("firewall missing")
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)}, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotFoundResponse(), errors.New("firewall missing")
+			},
+			expectedError: errFailedToCreateFirewallOnUpdate,
+		},
+		{
+			name:         "return error when failing to add inbound rules on update request to firewall API",
+			fwCache:      newFakeFirewallCache(diffFakeInboundRule),
+			inboundRules: []godo.InboundRule{fakeInboundRule},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
```

Comment:
@timoreimann we shouldn't hit the update call since `targetFirewall == nil` aka the firewall cache.

https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/6338bc26c26a4fefd5d61fe941a6bf7fb5756799/cloud-controller-manager/do/firewall_controller.go#L251

Reply:
Whoops, I think my comment was meant to reference line 463 where we mock out an error on the firewall service's `Get()` call. AFAIU, this test is about the following scenario:

- we do not have a firewall cached
- we invoke `firewallManager.Get()`
- `firewallManager.Get()` also realizes that we have nothing cached and does a `firewallService.List()`
- that `List()` call returns the stubbed out error

So `firewallService.Get()` should never be called and thus does not need to be stubbed out? Or am I missing something?